src/freezers.js

```diff
@@ -1,7 +1,42 @@
 import FrozenObject from './FrozenObject'
 import FrozenArray from './FrozenArray'
-/*
-  Create instances usint the constructors
-*/
-export const freezeObject = (o) => new FrozenObject(o)
-export const freezeArray = (a) => new FrozenArray(a)
+
+const isPlainObj = (value) => {
+  return value && (value.constructor === Object || value.constructor === undefined)
+}
+
+export const deepFreeze = (source) => {
+  return recursiveFreeze(source)
+}
+
+export const freeze = (source) => {
+  if (isPlainObj(source)) {
+    return new FrozenObject(source)
+  } else if (Array.isArray(source)) {
+    return new FrozenArray(source)
+  } else {
+    return source
+  }
+}
+
+const recursiveFreeze = (source, key) => {
+  let o = {}
+  if (isPlainObj(source)) {
+    Object.assign(o, keyIterator(source, (value, key) => recursiveFreeze(value, key)))
+  } else {
+    o[key] = source
+  }
+
+  if (Array.isArray(source)) {
+    return source.map(value => recursiveFreeze(value))
+  }
+
+  return freeze(o)
```

Comment:
I didn't find the way without having the `o` as a temp. I accomplish with freezing the childs of the source or at least the parent, not both. I don't know if there's a way more functional.